features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/browsers/OnmsVaadinContainer.java

```diff
@@ -173,15 +177,22 @@ public K getItemId(int rowIndex) {
         }
 
         public void reload(Page page) {
-            reset();
-            List<T> beans = getItemsForCache(m_datasource, page);
-            if (beans == null) return;
-            int rowNumber = page.getStart();
-            for (T eachBean : beans) {
-                addItem(rowNumber, getId(eachBean), eachBean);
-                doItemAddedCallBack(rowNumber, getId(eachBean), eachBean);
-                rowNumber++;
-            }
+            // Wrap the reload in the single transaction context to ensure
+            // that all of the callbacks are executed within a session
+            m_transactionTemplate.execute(new TransactionCallbackWithoutResult() {
```

Comment:
I thought about this myself.
What about moving this to the `OnmsDaoContainerDatasource` and remove the `doItemAddedCallback` method alltogether, as I think that logic should live in `OnmsDaoContainerDatasource` too.